packages/big-design/src/components/Worksheet/hooks/useCopyPaste/useCopyPaste.ts

```diff
@@ -0,0 +1,118 @@
+import { useCallback, useEffect, useMemo } from 'react';
+
+import { Cell, InternalWorksheetColumn } from '../../types';
+import { useUpdateItems } from '../useUpdateItems';
+import { useWorksheetStore } from '../useWorksheetStore';
+
+const isPasteAllowed = (
+  sourceCell: Cell<any>,
+  targetCell: Cell<any>,
+  columns: InternalWorksheetColumn<any>,
+) => {
+  const sourceType = sourceCell.type;
+  const targetType = targetCell.type;
+
+  return (
+    (!columns.disabled &&
+      sourceType === targetType &&
+      (sourceType === 'text' ||
+        sourceType === 'number' ||
+        sourceType === 'checkbox' ||
+        sourceCell.columnIndex === targetCell.columnIndex)) ||
+    (sourceType === 'text' && targetType === 'number') ||
+    (sourceType === 'number' && targetType === 'text')
+  );
+};
+
+export const useCopyPasteHandler = () => {
+  const { store, useStore } = useWorksheetStore();
+  const { updateItems } = useUpdateItems();
+
+  const selectedCells = useStore(
+    store,
+    useMemo(() => (state) => state.selectedCells, []),
+  );
+
+  const copiedCells = useStore(
+    store,
+    useMemo(() => (state) => state.copiedCells, []),
+  );
+
+  const rows = useStore(
+    store,
+    useMemo(() => (state) => state.rows, []),
+  );
+
+  const columns = useStore(
+    store,
+    useMemo(() => (state) => state.columns, []),
+  );
+
+  const disabledRows = useStore(
+    store,
+    useMemo(() => (state) => state.disabledRows, []),
+  );
+
+  const setCopiedCells = useStore(store, (state) => state.setCopiedCells);
+
+  const cellsToUpdate = useMemo(() => {
+    if (selectedCells.length === 0) {
+      return [];
+    }
+
+    const pasteFromIdx = selectedCells[0].rowIndex;
+    const cellIdxsToUpdate = Array.from(
+      { length: Math.min(copiedCells.length, rows.length - pasteFromIdx) },
+      (_, index) => pasteFromIdx + index,
+    );
+
+    return cellIdxsToUpdate.map((idxToUpdate) => ({
+      rowIndex: idxToUpdate,
+      columnIndex: selectedCells[0].columnIndex,
+      disabled: selectedCells[0].disabled,
+      hash: selectedCells[0].hash,
+      type: selectedCells[0].type,
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+      value: rows[idxToUpdate][selectedCells[0].hash],
```

Comment:
🍹 I'd see if we can remove this eslint-disable.

Reply:
Done)